cmd/src/actions_exec.go

```diff
@@ -0,0 +1,452 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+	"unicode/utf8"
+
+	"github.com/fatih/color"
+	"github.com/gosuri/uilive"
+	"github.com/pkg/errors"
+	"github.com/sourcegraph/go-diff/diff"
+)
+
+type ActionFile struct {
+	ScopeQuery string           `json:"scopeQuery,omitempty"`
+	Run        []*ActionFileRun `json:"run"`
+}
+
+type ActionFileRun struct {
+	Type               string   `json:"type"` // "command"
+	Dockerfile         string   `json:"dockerfile,omitempty"`
+	Image              string   `json:"image,omitempty"` // Docker image
+	ImageContentDigest string   // TODO!(sqs): do not allow setting
+	CacheDirs          []string `json:"cacheDirs,omitempty"`
+	Args               []string `json:"args,omitempty"`
+}
+
+type CampaignPlanPatch struct {
+	Repository   string `json:"repository"`
+	BaseRevision string `json:"baseRevision"`
+	Patch        string `json:"patch"`
+}
+
+func userCacheDir() (string, error) {
+	userCacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(userCacheDir, "sourcegraph-src"), nil
+}
+
+func init() {
+	usage := `
+Execute an action on code in repositories. The output of an action is a set of patches that can be used to create a campaign to open changesets and perform large-scale code changes.
+
+Examples:
+
+  Execute an action:
+
+    	$ src actions exec TODO
+
+`
+
+	flagSet := flag.NewFlagSet("exec", flag.ExitOnError)
+	usageFunc := func() {
+		fmt.Fprintf(flag.CommandLine.Output(), "Usage of 'src actions %s':\n", flagSet.Name())
+		flagSet.PrintDefaults()
+		fmt.Println(usage)
+	}
+	const stdin = "<stdin>"
+	userCacheDir, _ := userCacheDir()
+	if userCacheDir != "" {
+		userCacheDir = filepath.Join(userCacheDir, "action-exec")
+	}
+	displayUserCacheDir := strings.Replace(userCacheDir, os.Getenv("HOME"), "$HOME", 1)
+	var (
+		fileFlag        = flagSet.String("f", "<stdin>", `The action file. (required)`)
+		parallelismFlag = flagSet.Int("j", runtime.GOMAXPROCS(0), "The number of parallel jobs.")
+		cacheDirFlag    = flagSet.String("cache", displayUserCacheDir, `Directory for caching results.`)
+		keepLogsFlag    = flagSet.Bool("keep-logs", false, "Do not remove execution log files when done.")
+	)
+
+	handler := func(args []string) error {
+		flagSet.Parse(args)
+
+		if *cacheDirFlag == displayUserCacheDir {
+			*cacheDirFlag = userCacheDir
+		}
+
+		var (
+			actionFile []byte
+			err        error
+		)
+		if *fileFlag == stdin {
+			// TODO!(sqs) easy to forget -f flag and have it hang forever
+			actionFile, err = ioutil.ReadAll(os.Stdin)
```

Comment:
If we read from `stdin`, we [can check whether it's a pipe](https://coderwall.com/p/zyxyeg/golang-having-fun-with-os-stdin-and-shell-pipes). If not, we exit with an error.